### DISCUSSION TITLE
How to convert RAW12 to any common format(tiff, jpeg, etc.) using ImageMagick

### DISCUSSION THREAD

**Discussion by gennadyg** [ImageMagick/ImageMagick#3109] [Help]
Hi,
Trying to use a following command to convert RAW12 image to PNG, which worked, but received black/white image, what parameters should be used to receive colors ? 

`cat IT8-chart-15ms.raw12 | convert \( -size 4096x3072 -depth 12 gray:- \) \( -clone 0 -crop -1-1 \) \( -clone 0 -crop -1+0 \) \( -clone 0 -crop +0-1 \) -sample 2048x1536 \( -clone 2,3 -average \) -delete 2,3 -swap 0,1 +swap -combine IT8-chart-15ms.png`

[RAW12 Image](http://files.apertus.org/AXIOM-Beta/snapshots/IT8-chart-15ms.raw12)
[Image that want to achieve](http://files.apertus.org/AXIOM-Beta/snapshots/IT8-chart-15ms_colorpreview.png)

Any help is appreciated.

**Comment by snibgo** [comment]
What version of IM, on what platform?

What is a "RAW12" image? Do you have a reference for this format?

After your "-sample" operation, the four images are identical. After that, no operation makes them non-identical. So, when you "-combine" them, the result is a gray image.

**Reply by gennadyg** [reply]
Thanks for checking, RAW12 description is [here](https://wiki.apertus.org/index.php/RAW12), appreciate your help.

**Reply by hahecd** [reply]
I refer to [this](https://www.1stvision.com/cameras/IDS/IDS-manuals/en/basics-raw-bayer-pixel-formats.html) for RAW12, here it states the pixel of RAW12 will take up 2 bytes anyway with 4 padding bits.

**Comment by snibgo** [comment]
Okay. This is a mosaic image, probably made by a digital camera, in the common RGGB pattern. From my [Demosaicing](http://im.snibgo.com/demosaic.htm) page, here is a primitive method to demosaic the image. Windows BAT syntax:
```
magick ^
  -size 4096x3072 -depth 12 gray:IT8-chart-15ms.raw12 ^
  -depth 16 ^
  ( -clone 0 -define sample:offset=25 -sample 50%% -resize 200%% ) ^
  ( -clone 0 ^
    ( -clone 0 -define sample:offset=75x25 -sample 50%% -resize 200%% ) ^
    ( -clone 0 -define sample:offset=25x75 -sample 50%% -resize 200%% ) ^
    -delete 0 ^
    -evaluate-sequence mean ^
  ) ^
  ( -clone 0 -define sample:offset=75 -sample 50%% -resize 200%% ) ^
  -delete 0 ^
  -combine ^
  -auto-level ^
  -resize 600 ^
  out.jpg
```
I "-auto-level" to stretch the values from 0 to 100% of QuantumRange, "-resize" to 600, and write to JPG for the web. In general, there is no need for the final resize, and the image should be saved to TIFF or PNG or whatever you want.
![out](https://user-images.githubusercontent.com/33812368/104314592-9040dc80-54d1-11eb-9777-a0d87d6629ed.jpg)

Another possibility is to write the mosaic image to a TIFF file, with extra metadata that would make it a DNG file, and that DNG file could be demosaiced by dcraw, or ImageMagick using libraw, or any number of other raw processor programs.

EDIT: Your "Image that want to achieve" is half size. If that is what you want, then remove each "-resize 200%%".

**Reply by gennadyg** [reply]
Thanks for your quick answer, looks great, currently OOO, will verify on my environment first thing tomorrow morning.

**Reply by gennadyg** [reply]
All worked for me, thank you very much.

**Comment by hahecd** [comment]
Hi,

I have a 1920x1542 mipi raw12 file with format of SBGGR12 wanting to achieve the same, when I tried the first command. it gave me the below error:

$ cat 1920_1542.raw12 | convert \( -size 1920x1542 -depth 12 gray:- \) \( -clone 0 -crop -1-1 \) \( -clone 0 -crop -1+0 \) \( -clone 0 -crop +0-1 \) -sample 1920x1542 \( -clone 2,3 -average \) -delete 2,3 -swap 0,1 +swap -combine 1920_1542.png
convert-im6.q16: unexpected end-of-file `-': No such file or directory @ error/gray.c/ReadGRAYImage/240.
convert-im6.q16: image sequence is required `-clone' @ error/convert.c/ConvertImageCommand/948.

can you help here?

Thanks,
Allen

**Comment by snibgo** [comment]
> unexpected end-of-file

That tells you the input file doesn't have enough data. If the file has 1920x1542 pixels, one channel, at 12 bits per pixel, then it should have 1920\*1542\*12 = 35527680 bits which is 4440960 bytes.

EDIT: corrected formatting.

**Comment by hahecd** [comment]
the raw file is 5921280 bytes, which is the result of 1920x1542x2. that means each pixel would take up 2 bytes. isn't this true for RAW12 data?
I used the PixelViewer to render the data before but I can only get the monochrome picture as below:
![picture](https://github.com/user-attachments/assets/8bc0f6c7-2d53-4b21-a98a-ff0df31609f1)
 
can you please help produce the picture in color in your way? attached please find the raw data file.
[fubba2.zip](https://github.com/user-attachments/files/16675584/fubba2.zip)

**Comment by snibgo** [comment]
> I have a 1920x1542 mipi raw12 file with format of SBGGR12 ...

I don't know the meaning of "mipi" or "SBGGR12".

Assuming the Bayer pattern is:
```
  B G
  G R
```
... and there are 12 bits per sample, not packed, then:
```
%IMG7%magick ^
  -size 1920x1542 ^
  -depth 16 ^
  gray:fubba2.raw ^
  -evaluate Multiply %%[fx:QuantumRange/4095] ^
  ( -clone 0 -define sample:offset=25 -sample 50%% -resize 200%% ) ^
  ( -clone 0 ^
    ( -clone 0 -define sample:offset=75x25 -sample 50%% -resize 200%% ) ^
    ( -clone 0 -define sample:offset=25x75 -sample 50%% -resize 200%% ) ^
    -delete 0 ^
    -evaluate-sequence mean ^
  ) ^
  ( -clone 0 -define sample:offset=75 -sample 50%% -resize 200%% ) ^
  -delete 0 ^
  -swap 0,2 ^
  -combine ^
  -resize 600 -quality 40 x.jpg
```
The result is:

![x](https://github.com/user-attachments/assets/8e6d5720-c8f3-4459-87cd-16ab394c2fe4)

**Reply by hahecd** [reply]
Hi,

Can you please check [this](https://www.1stvision.com/cameras/IDS/IDS-manuals/en/basics-raw-bayer-pixel-formats.html) for information about RAW12?
My sensor is producing the raw data with the below RAW Bayer pattern:
![image](https://github.com/user-attachments/assets/0f989cd7-c7f7-49ea-9777-d45bf1c8a091)

And I would expect some picture similar to below:
![colored](https://github.com/user-attachments/assets/d9a71153-27f3-4a27-970d-49e34ab588b3)

**Comment by snibgo** [comment]
The result from my command above is similar to your expected result.

Instead of that command, you could use the `bayer:` format. That assumes the CFA is RGGB but your image is BGGR so we need to swap the red and blue channels.
```
magick ^
  -size 1920x1542 ^
  bayer:fubba2.raw ^
  -evaluate Multiply %%[fx:QuantumRange/4095] ^
  -separate -swap 0,2 -combine ^
  out.png
```

My command uses a very simple and primitive debayering method. More sophisticated methods are available in programs such as dcraw, darktable and RawTherapee.

My command doesn't include arithmetic on the three channels, a "camera matrix", that would adjust hue and saturation to some standard colour primaries (including a white balance). Nor does it apply a tone transfer curve. Those operations would transform the image to sRGB or other colorspace.

The specification for your camera may give the details of these transformations.

If you use the camera to photograph a colour chart, then use another known camera or scanner on the same colour chart, a colour matrix that transforms from the first image to the second can be calculated. I show how to do this at [Colour checker charts](http://im.snibgo.com/colchkcha.htm). Programs such as darktable can also do this.

**Reply by hahecd** [reply]
thank you for the details!
can you please provide the equivalent linux command line for the above "magick" commands?

**Comment by snibgo** [comment]
To convert these from Windows BAT syntax to bash syntax, change double `%%` to single `%`, change caret `^` to backslash `\`, and escape each parenthesis with backslash `\(` and `\)`.
```
magick \
  -size 1920x1542 \
  -depth 16 \
  gray:fubba2.raw \
  -evaluate Multiply %[fx:QuantumRange/4095] \
  \( -clone 0 -define sample:offset=25 -sample 50% -resize 200% \) \
  \( -clone 0 \
    \( -clone 0 -define sample:offset=75x25 -sample 50% -resize 200% \) \
    \( -clone 0 -define sample:offset=25x75 -sample 50% -resize 200% \) \
    -delete 0 \
    -evaluate-sequence mean \
  \) \
  \( -clone 0 -define sample:offset=75 -sample 50% -resize 200% \) \
  -delete 0 \
  -swap 0,2 \
  -combine \
  -resize 600 -quality 40 x.jpg

magick \
  -size 1920x1542 \
  bayer:fubba2.raw \
  -evaluate Multiply %[fx:QuantumRange/4095] \
  -separate -swap 0,2 -combine \
  -resize 600 -quality 40 x2.jpg
```
EDIT: Instead of escaping parentheses, some people prefer to quote them: `"("` and `")"`.

**Reply by hahecd** [reply]
the commands can run into success. but result files can not be opened with any picture viewers in Windows.
![image](https://github.com/user-attachments/assets/86552249-2d3f-4f68-884f-124b3e656811)

**Comment by snibgo** [comment]
> the commands can run into success. but result files can not be opened with any picture viewers in Windows.

All the commands I have shown work for me.

Please try this:
```
magick rose: x.jpg
```
Can you view x.jpg?

What is your platform (Windows BAT, or bash or what)? What version of IM? What is the text output from `magick -version`? For me, this is:
```
Version: ImageMagick 7.1.1-20 Q16-HDRI x86 98bb1d4:20231008 https://imagemagick.org
Copyright: (C) 1999 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI OpenCL OpenMP(2.0)
Delegates (built-in): bzlib cairo freetype gslib heic jng jp2 jpeg jxl lcms lqr lzma openexr pangocairo png ps raqm raw rsvg tiff webp xml zip zlib
Compiler: Visual Studio 2022 (193532217)
```
Did you use the fubba2.raw file already posted, or some other input? If some other, please zip it and upload that here. What command did you use? Please zip your output and upload that here.

**Reply by hahecd** [reply]
$ magick rose: x.jpg
$
nothing happened when doing this

$ magick --version
Version: ImageMagick 7.1.1-36 Q16-HDRI x86_64 22352 https://imagemagick.org
Copyright: (C) 1999 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI OpenMP(4.5)
Delegates (built-in):
Compiler: gcc (11.4)

$ uname -a
Linux ctu-shared-dev2 6.5.0-45-generic #45~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Mon Jul 15 16:40:02 UTC 2 x86_64 x86_64 x86_64 GNU/Linux
$

command issued:
$ magick -size 1920x1542 -depth 16 gray:fubba.raw12 -evaluate Multiply %[fx:QuantumRange/4095] \( -clone 0 -define sample:offset=25 -sample 50% -resize 200% \) \( -clone 0 \( -clone 0 -define sample:offset=75x25 -sample 50% -resize 200% \) \( -clone 0 -define sample:offset=25x75 -sample 50% -resize 200% \) -delete 0 -evaluate-sequence mean \) \( -clone 0 -define sample:offset=75 -sample 50% -resize 200% \) -delete 0 -swap 0,2 -combine -resize 600 -quality 40 x.jpg


fubba.raw12 is the same file as the posted fubba2.raw file. I attached it here anyway.
[fubba.zip](https://github.com/user-attachments/files/16729084/fubba.zip)

**Comment by snibgo** [comment]
> Delegates (built-in):

That's the problem. You have no built-in delegates, which means your IM can't write to PNG, JPG, TIFF or anything much. Compare that to the list on my computer:
```
Delegates (built-in): bzlib cairo freetype gslib heic jng jp2 jpeg jxl lcms lqr lzma openexr pangocairo png ps raqm raw rsvg tiff webp xml zip zlib
```
You need to download delegates and re-build IM. I don't use Linux, and can't advise.

**Comment by fmw42** [comment]
Perhaps this will help.

https://askubuntu.com/questions/1042436/how-to-install-delegate-libraries-for-image-magick-7-0-7

**Reply by hahecd** [reply]
This truly helps. thank you!
